### ISSUE TITLE
Versioning

### ISSUE THREAD

**Issue by david4096** [ga4gh/data-repository-service-schemas#9]
We need to better constrain how versioning is expected to be managed in practice.

Via @bwalsh

> How exactly do we have a version? The what was the timestamp of the previous version? What was its hash value? Should we probably have a separate DataObject repeated versions to hold different versions?

https://cloud.google.com/storage/docs/key-terms#immutability
http://docs.aws.amazon.com/AmazonS3/latest/dev/ObjectVersioning.html

> Google in particular has a version for object and a metaversion for metadata changes.
https://cloud.google.com/storage/docs/object-versioning#details
> This differs from amazon:
> Each Amazon S3 object has data, a key, and metadata. Object key (or key name) uniquely identifies the object in a bucket. Object metadata is a set of name-value pairs. You can set object metadata at the time you upload it. After you upload the object, you cannot modify object metadata. The only way to modify object metadata is to make a copy of the object and set the metadata.

> S3 breaks meta data into two arrays, system and user
http://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html#object-metadata

**Comment by philloooo** [comment]
this is our tentative plan for indexd https://github.com/LabAdvComp/indexd/issues/45

**Comment by philloooo** [comment]
just to take notes from our f2f conversation...
We agreed system identifier(UUID) -> hashes should be immutable, one identifier always point to the same piece of data.
The question is should an alias always point to the same piece of data, or it can be mutable, and point to the latest/all version of data.
optionally we could support both alias to ConceptObject and alias to DataObject.
Relational model:
```
ConceptAlias {
  identifier
  concept_id
  some_metadata
}
DataAlias {
  identifier
  []hashes
  some_metadata
}
ConceptObject {
  identifier
  some_metadata
}
DataObject {
  identifier
  concept_id
  version
  []hashes
  []urls
}
```
compact document:
```
ConceptObject {
  identifier
  []aliases
  []DataObjects {
    identifier
    version
    []urls
    []hashes
    []aliases
  }
}

**Comment by david4096** [comment]
Addressed in #24 to support a variety of versioning requirements. Leaving open until this data model is proven against two different services that support versioning.
